qiita_pet/handlers/download.py

```diff
@@ -143,5 +143,87 @@ def get(self, extras):
                         '/protected-working_dir/' + relpath)
         self.set_header('Content-Disposition',
                         'attachment; filename=%s' % basename(relpath))
+        self.finish()
+
+
+class DownloadRawData(BaseHandler):
+    @authenticated
+    @execute_as_transaction
+    def get(self, study_id):
+        study_id = int(study_id)
+        # Check general access to study
+        study_info = study_get_req(study_id, self.current_user.id)
+        if study_info['status'] != 'success':
+            raise HTTPError(405, "%s: %s, %s" % (study_info['message'],
+                                                 self.current_user.email,
+                                                 str(study_id)))
+
+        study = Study(study_id)
+        user = self.current_user
+        # Check "owner" access to the study
+        if not study.has_access(user, True):
+            raise HTTPError(405, "%s: %s, %s" % ('No raw data access',
+                                                 self.current_user.email,
+                                                 str(study_id)))
+
+        basedir = get_db_files_base_dir()
+        basedir_len = len(basedir) + 1
+        # loop over artifacts and retrieve raw data (no parents)
+        to_download = []
+        vfabu = validate_filepath_access_by_user
+        for a in study.artifacts():
+            if not a.parents:
+                to_add = True
+                for i, (fid, path, data_type) in enumerate(a.filepaths):
+                    # validate access only of the first artifact filepath,
+                    # the rest have the same permissions
+                    if (i == 0 and not vfabu(user, fid)):
```

Comment:
To clarify, validate_filepath_access_by_user checks that the specific user has access to the specific filepath id based on it's data type so we are checking that the user has access to that filepath and if it does it allows download to all the other filepaths within that artifact. Remember an artifact can be formed by several files (filepath ids).